server/controllers/events/events_controller.go

```diff
@@ -44,21 +50,116 @@ const bitbucketCloudRequestIDHeader = "X-Request-UUID"
 const bitbucketServerRequestIDHeader = "X-Request-ID"
 const bitbucketServerSignatureHeader = "X-Hub-Signature"
 
+func NewRequestResolvers(
+	providerResolverInitializer map[models.VCSHostType]func() RequestResolver,
+	supportedProviders []models.VCSHostType,
+) []RequestResolver {
+	var resolvers []RequestResolver
+	for provider, resolverInitializer := range providerResolverInitializer {
+		for _, supportedProvider := range supportedProviders {
+
+			if provider != supportedProvider {
+				continue
+			}
+
+			resolvers = append(resolvers, resolverInitializer())
+		}
+	}
+
+	return resolvers
+}
+
+func NewVCSEventsController(
+	legacyLogger logging.SimpleLogging,
+	scope tally.Scope,
+	githubWebhookSecret []byte,
+	allowDraftPRs bool,
+	commandRunner events.CommandRunner,
+	commentParser events.CommentParsing,
+	eventParser events.EventParsing,
+	pullCleaner events.PullCleaner,
+	repoAllowlistChecker *events.RepoAllowlistChecker,
+	vcsClient vcs.Client,
+	logger logging.Logger,
+	applyDisabled bool,
+	gitlabWebhookSecret []byte,
+	supportedVCSProviders []models.VCSHostType,
+	bitbucketWebhookSecret []byte,
+	azureDevopsWebhookBasicUser []byte,
+	azureDevopsWebhookBasicPassword []byte,
+	repoConverter github_converters.RepoConverter,
+	pullConverter github_converters.PullConverter,
+) *VCSEventsController {
+
+	prHandler := handlers.NewPullRequestEvent(
+		repoAllowlistChecker, vcsClient, pullCleaner, commandRunner,
+	)
+
+	commentHandler := handlers.NewCommentEvent(
+		commentParser,
+		repoAllowlistChecker,
+		vcsClient,
+		commandRunner,
+	)
+
+	// lazy map of resolver providers to their resolver
+	// laziness ensures we only instantiate the providers we support.
+	providerResolverInitializer := map[models.VCSHostType]func() RequestResolver{
```

Comment:
curious why not do pass this lazy initializer as an argument? You can modify the lazy function to accept context specific arguments like `prHandler` and `commentHandler`? Probably can also create the map based on supported `VCSProviders` and don't have to do filtering in the `NewRequestResolvers`

Reply:
mainly did this for simplicity and for keeping related initializers in their respective packages. (I can't stand adding to server.go)

that being said, not sure I completely follow what you're suggesting, mind writing some pseduocode as an example

Reply:
Yep, here is what I'm thinking. We can change `providerResolverInitializer` to something like this
```golang
map[models.VCSHostType]func() events_controllers.RequestResolver{
               models.Github: func(prHandler *handlers.PullRequestEvent, commentHandler *handlers.CommentEvent) events_controllers.RequestResolver {
			return github.NewRequestHandler(
				legacyLogger,
				scope,
				webhookSecret,
				commentHandler,
				prHandler,
				allowDraftPRs,
				repoConverter,
				pullConverter,
			)
		},
}
```
And in the initializer can just loop through it and pass in `prHandler`, and `commentHandler` as an argument to the lazy function. 
As for the second of my comment, instead of passing in `supportedVCSHosts`, we can create and populate `providerResolverInitializer` on the server and pass that in. But i realize that might take you for another rabbit hole of refactoring with changing how we initialize our vcs clients.

Reply:
> Yep, here is what I'm thinking. We can change providerResolverInitializer to something like this

I think this gets a bit too specific though, we run the risk of needing to modify the interface to support more context specific members



> As for the second of my comment, instead of passing in supportedVCSHosts, we can create and populate providerResolverInitializer on the server and pass that in. But i realize that might take you for another rabbit hole of refactoring with changing how we initialize our vcs clients.

Yeah supportedhosts is still required for the other vcs providers.